crates/huub/src/constraints/int_value_precede.rs

```diff
@@ -0,0 +1,1144 @@
+		}
+		//todo this can become more powerful if updated upper bound from previous loop is available
+		self.vars.retain(|&var| {
+			self.values
+				.iter()
+				.any(|&val| actions.check_int_in_domain(var, val))
+		});
+		if self.vars.is_empty() {
+			return Ok(SimplificationStatus::Subsumed);
+		}
+		Ok(SimplificationStatus::Fixpoint)
+	}
+
+	fn to_solver(&self, slv: &mut dyn ReformulationActions) -> Result<(), ReformulationError> {
+		let vars: Vec<_> = self.vars.iter().map(|v| slv.get_solver_int(*v)).collect();
+		IntValuePrecedeChainValue::new_in(slv, self.values.clone(), vars);
+		Ok(())
+	}
+}
+
+impl IntValuePrecedeChainValue {
+	/// 3 cases:
+	/// - Current lower bound index is above k - This is the value that required
+	///   the earlier lower bound.
+	/// - Index k is in the domain of var i - Go one step up and to the next
+	///   variable.
+	/// - Index k is not in the domain of var i - i can be anything else, go to
+	///   next variable.
+	fn ex_l<P: PropagationActions>(&self, actions: &mut P, i: usize, j: usize) -> Vec<BoolView> {
+		if let Some(lb) = self.get_lower_bound(actions, i) {
+			if lb > j {
+				let mut v = vec![
+					actions.get_int_lit(self.vars[i], IntLitMeaning::GreaterEq(self.min_val)),
+					actions.get_int_lit(self.vars[i], IntLitMeaning::Less(self.max_val + 1)),
+				];
+				v.append(
+					&mut self
+						.holes
+						.iter()
+						.map(|&h| actions.get_int_lit(self.vars[i], IntLitMeaning::NotEq(h)))
+						.collect(),
+				);
+				v.append(
+					&mut (0..j)
+						.map(|k| {
+							actions.get_int_lit(self.vars[i], IntLitMeaning::NotEq(self.values[k]))
+						})
+						.collect(),
+				);
+				return v;
+			}
+		}
+		if actions.check_int_in_domain(self.vars[i], self.values[j - 1]) {
+			return self.ex_l(actions, i + 1, j + 1);
+		}
+		let mut v = self.ex_l(actions, i + 1, j);
+		v.push(actions.get_int_lit(self.vars[i], IntLitMeaning::NotEq(self.values[j - 1])));
+		v
+	}
+
+	/// Lower bound explanation: Could not have this index earlier (=upper bound
+	/// explanation) and some later index requires the lower bound.
+	fn explain_lower<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		j: usize,
+	) -> Vec<BoolView> {
+		let mut v = self.ex_l(actions, i + 1, j);
+		if j > 0 {
+			v.append(&mut self.explain_upper(actions, i, j));
+		}
+		v
+	}
+
+	/// Upper bound explanation: All previous indices are smaller.
+	fn explain_upper<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		j: usize,
+	) -> Vec<BoolView> {
+		self.vars
+			.iter()
+			.take(i)
+			.map(|&v| actions.get_int_lit(v, IntLitMeaning::NotEq(self.values[j - 1])))
+			.collect()
+	}
+
+	/// Get the lower bound for the index in values, None if any options outside
+	/// values are still in the domain. Has to exclude values below and above the
+	/// range of values, then all holes, finally values with lower index.
+	fn get_lower_bound<I: InspectionActions>(&self, actions: &mut I, i: usize) -> Option<usize> {
```

Comment:
The naming of this method is a little confusing with `get_int_lower_bound`. Maybe we can think of something more clear.